indexer/src/main/java/bio/terra/tanagra/indexing/job/dataflow/CleanHierarchyNodesWithZeroCounts.java

```diff
@@ -0,0 +1,102 @@
+package bio.terra.tanagra.indexing.job.dataflow;
+
+import bio.terra.tanagra.api.shared.BinaryOperator;
+import bio.terra.tanagra.indexing.job.BigQueryJob;
+import bio.terra.tanagra.query.bigquery.translator.BQApiTranslator;
+import bio.terra.tanagra.query.sql.SqlField;
+import bio.terra.tanagra.query.sql.SqlQueryField;
+import bio.terra.tanagra.underlay.entitymodel.Hierarchy;
+import bio.terra.tanagra.underlay.entitymodel.entitygroup.EntityGroup;
+import bio.terra.tanagra.underlay.indextable.ITEntityMain;
+import bio.terra.tanagra.underlay.serialization.SZIndexer;
+import com.google.cloud.bigquery.Table;
+import com.google.cloud.bigquery.TableResult;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CleanHierarchyNodesWithZeroCounts extends BigQueryJob {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(WriteRollupCounts.class);
+
+  private final EntityGroup entityGroup;
+  private final ITEntityMain indexTable;
+  private final Hierarchy hierarchy;
+
+  public CleanHierarchyNodesWithZeroCounts(
+      SZIndexer indexerConfig,
+      EntityGroup entityGroup,
+      ITEntityMain indexTable,
+      Hierarchy hierarchy) {
+    super(indexerConfig);
+    this.entityGroup = entityGroup;
+    this.indexTable = indexTable;
+    this.hierarchy = hierarchy;
+  }
+
+  @Override
+  public String getEntityGroup() {
+    return entityGroup.getName();
+  }
+
+  @Override
+  protected String getOutputTableName() {
+    return indexTable.getTablePointer().getTableName();
+  }
+
+  @Override
+  protected Optional<Table> getOutputTable() {
+    return googleBigQuery.getTable(
+        indexerConfig.bigQuery.indexData.projectId,
+        indexerConfig.bigQuery.indexData.datasetId,
+        getOutputTableName());
+  }
+
+  @Override
+  public JobStatus checkStatus() {
+    return getOutputTable().isPresent() && outputTableHasNoNodesWithZeroCounts()
+        ? JobStatus.COMPLETE
+        : JobStatus.NOT_STARTED;
+  }
+
+  @Override
+  public void run(boolean isDryRun) {
```

Comment:
I think you also need to update the `num_children` field for the nodes that remain, since those counts were calculated before this job. Currently that field is used by the UI to determine whether to show an expander icon in the hierarchy view.

Reply:
I'm not sure of an easy way to accomplish updating num_children. 
Will have to do the following:
- Delete from T_HAD_xxx, T_HCP_xxx and T_PNC_xxx
- rerun WriteChildParent
- rerun WriteAncestorDescendant
- rerun WriteNumChildrenAndPaths - dependent on WriteChildParent and WriteAncestorDescendant

This is essentially rerunning the entire job. Not sure I want to have to build this twice for each hierarchy. May need to take another approach.

Reply:
I don't think you need to re-run all those jobs.

- The child-parent and ancestor-descendant tables should still be correct, though they'll contain some extra relationships now. You could delete those, but I don't think that will affect anything.
- The num_children may be incorrect, but that might be possible to update with just a SQL query, without re-running the dataflow job.
- The path you might also be okay without updating -- there could certainly be paths that include nodes you deleted, but if you're doing this based on rollup count, I think the only nodes that could contain deleted nodes would themselves be deleted.

Not sure I'm considering all the cases, so happy to discuss further if you think that doesn't check out.

Reply:
This looks great, thanks. Since you updated the child-parent table, could you also update the ancestor-descendant table, just for consistency please? I think it's basically an identical function to `cleanChildParent`, but use the `ITHierarchyAncestorDescendant` table instead.